### PR TITLE
Broker recovery

### PR DESCRIPTION
**Pull request by benbjohnson**
## Overview

This pull request fixes the broker recovery so that it determines the last index from the various topic logs instead of persisting the snapshot on every message that comes in.

### CODE REVIEW

File: messaging/broker.go [hunk 1]
```diff
@@ -57,6 +57,14 @@ func (b *Broker) metaPath() string {
 	return filepath.Join(b.path, "meta")
 }
 
+// Index returns the highest index seen by the broker.
```
Comment (otoolep) [inline messaging/broker.go]: Minor: "seen" is a vague term. Could it be improved?

File: raft/log.go [hunk 2]
```diff
@@ -45,7 +45,7 @@ const logEntryHeaderSize = 8 + 8 + 8 // sz+index+term
 
 // WaitInterval represents the amount of time between checks to the applied index.
 // This is used by clients wanting to wait until a given index is processed.
-const WaitInterval = 100 * time.Millisecond
+const WaitInterval = 1 * time.Millisecond
```
Comment (otoolep) [inline raft/log.go]: Deliberate? Just want to be sure.

File: messaging/broker.go (hunk 1 above)
Comment (otoolep) [inline messaging/broker.go]: I guess somewhere we need to capture that the significance of this is that highest across all topics has been safely applied in the Raft log, if I remember correctly.

File: messaging/broker_test.go [hunk 3]
```diff
@@ -194,6 +194,31 @@ func TestBroker_Unsubscribe_ErrReplicaNotFound(t *testing.T) {
...
 }
 
+// Ensure the broker can reopen and recover correctly.
+func TestBroker_Reopen(t *testing.T) {
+	b := NewBroker(nil)
+	defer b.Close()
+	b.MustCreateReplica(2000, &url.URL{Host: "localhost"})
+	b.MustSubscribe(2000, 20)
+	b.MustPublishSync(&messaging.Message{TopicID: 20, Data: []byte("0000")})
+	b.MustPublishSync(&messaging.Message{TopicID: 20, Data: []byte("0000")})
```
Comment (otoolep) [inline messaging/broker_test.go]: To fully test the code, should this test publish to at least two different topics?

**Comment by otoolep** [comment]
Some suggested changes, but this code makes sense to me. I would like to see the test tweaked  

+1

File: raft/log.go (hunk 2 above)
Comment (otoolep) [inline raft/log.go]: Ok, you have an explicit commit for this, so all good.

File: messaging/broker.go [hunk 4]
```diff
@@ -774,6 +791,33 @@ func (t *topic) Close() error {
 	return nil
 }
 
+// loadIndex reads the highest available index for a topic from disk.
+func (t *topic) loadIndex() error {
```
Comment (pauldix) [inline messaging/broker.go]: This seems like it's going to take a long time if there are quite a bit of topic logs backed up.

Comment (benbjohnson) [inline messaging/broker.go]: When we add truncation (and log segments) then it'll only have to traverse through the latest log segment for each topic.

**Comment by pauldix** [comment]
LGTM, :shipit:

**Pull request merged**